huaweicloud/resource_huaweicloud_rds_instance_v3.go

```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
+		}
+		createRdsOpts.BackupStrategy = &backup
+	}
+	if hasFilledOpt(d, "ha_replication_mode") {
+		ha := instances.Ha{
+			Mode:            "ha",
+			ReplicationMode: d.Get("ha_replication_mode").(string),
+		}
+		createRdsOpts.Ha = &ha
 	}
 
-	timeout := d.Timeout(schema.TimeoutCreate)
-	obj, err := asyncWaitRdsInstanceV3Create(d, config, r, client, timeout)
+	res, err := instances.Create(client, createRdsOpts).Extract()
 	if err != nil {
 		return err
 	}
-	id, err := navigateValue(obj, []string{"job", "instance", "id"}, nil)
-	if err != nil {
-		return fmt.Errorf("Error constructing id, err=%s", err)
+	d.SetId(res.Instance.Id)
+
+	log.Printf("[DEBUG] Waiting for instance (%s) to become running", d.Id())
+	// After the rds instance is created, it will enter the backup state.
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"BUILD", "ACTIVE"},
+		Target:     []string{"BACKING UP"},
+		Refresh:    InstanceV3StateRefreshFunc(client, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+	if _, err = stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for instance (%s) to become backup: %s", d.Id(), err)
+	}
+	// Only when the backup is completed, the instance is considered to be created.
+	stateConf = &resource.StateChangeConf{
+		Pending:    []string{"BACKING UP"},
+		Target:     []string{"ACTIVE"},
+		Refresh:    InstanceV3StateRefreshFunc(client, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+	if _, err = stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for instance (%s) to become ready: %s", d.Id(), err)
 	}
-	instanceID := id.(string)
-	d.SetId(instanceID)
 
 	//set tags
 	tagRaw := d.Get("tags").(map[string]interface{})
 	if len(tagRaw) > 0 {
 		taglist := expandResourceTags(tagRaw)
-		if tagErr := tags.Create(client, "instances", instanceID, taglist).ExtractErr(); tagErr != nil {
-			return fmt.Errorf("Error setting tags of RDS instance %s: %s", instanceID, tagErr)
+		if tagErr := tags.Create(client, "instances", d.Id(), taglist).ExtractErr(); tagErr != nil {
+			return fmt.Errorf("Error setting tags of RDS instance %s: %s", d.Id(), tagErr)
 		}
 	}
 
 	return resourceRdsInstanceV3Read(d, meta)
 }
 
+func resourceRdsInstanceV3Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.RdsV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating sdk client, err=%s", err)
+	}
+	page, err := instances.List(client, instances.ListRdsInstanceOpts{Id: d.Id()}).AllPages()
+	if err != nil {
+		return fmt.Errorf("Error fetching rds instances %s from HuaweiCloud: %s", d.Id(), err)
+	}
+	rdsList, err := instances.ExtractRdsInstances(page.(instances.RdsPage))
+	if len(rdsList.Instances) != 1 {
+		return fmt.Errorf("The rds instance is not exist")
+	}
```

Comment:
you can use `getRdsInstanceByID`

Reply:
All rds List function are replaced with `getRdsInstanceByID` and use `reflect.DeepEqual` to check whether instance is empty.